signature/policy_eval_sigstore.go

```diff
@@ -184,43 +221,63 @@ func (pr *prSigstoreSigned) isSignatureAccepted(ctx context.Context, image priva
 		if err != nil {
 			return sarRejected, err
 		}
-		publicKey = pk
+		publicKeys = []crypto.PublicKey{pk}
 	}
 
-	if publicKey == nil {
+	if len(publicKeys) == 0 {
 		// Coverage: This should never happen, we have already excluded the possibility in the switch above.
 		return sarRejected, fmt.Errorf("Internal inconsistency: publicKey not set before verifying sigstore payload")
 	}
-	signature, err := internal.VerifySigstorePayload(publicKey, untrustedPayload, untrustedBase64Signature, internal.SigstorePayloadAcceptanceRules{
-		ValidateSignedDockerReference: func(ref string) error {
-			if !pr.SignedIdentity.matchesDockerReference(image, ref) {
-				return PolicyRequirementError(fmt.Sprintf("Signature for identity %q is not accepted", ref))
-			}
-			return nil
-		},
-		ValidateSignedDockerManifestDigest: func(digest digest.Digest) error {
-			m, _, err := image.Manifest(ctx)
-			if err != nil {
-				return err
-			}
-			digestMatches, err := manifest.MatchesDigest(m, digest)
-			if err != nil {
-				return err
-			}
-			if !digestMatches {
-				return PolicyRequirementError(fmt.Sprintf("Signature for digest %s does not match", digest))
-			}
-			return nil
-		},
-	})
-	if err != nil {
-		return sarRejected, err
-	}
-	if signature == nil { // A paranoid sanity check that VerifySigstorePayload has returned consistent values
-		return sarRejected, errors.New("internal error: VerifySigstorePayload succeeded but returned no data") // Coverage: This should never happen.
+
+	errs := make([]error, len(publicKeys))
+	hasPolicyRequirementError := false
+
+	for _, publicKey := range publicKeys {
+		signature, err := internal.VerifySigstorePayload(publicKey, untrustedPayload, untrustedBase64Signature, internal.SigstorePayloadAcceptanceRules{
```

Comment:
I think it makes more sense to provide all `publicKeys` to a single call to `VerifySigstorePayload`.

The `untrustedPayload` and `untrustedBase64Signature` are the same in all calls, so basically there are three cases:
- There’s something wrong about the basic format (e.g. incorrect base64): we should only report that once, not N times for N keys
- The basic format is fine, but the signature is not valid under any of the keys: fine, probably N failures? Although the error messages returned seem not to be all that informative, we might just want to consolidate
- The cryptographic signature does validate under at least one key, but the payload is rejected by `SigstorePayloadAcceptanceRules`: we should only report the payload error, and not mention anything about failing to validate with the other public keys.

All of that seems easier to do if the loop is inside `VerifySigstorePayload` rather than outside.

Reply:
This is a very good idea, it simplified the code quite a lot